File: study_tracker_flutter/lib/data/repositories/ai_feature_settings_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:drift/drift.dart' as drift;
import '../db/database.dart' as db;
import '../db/database_provider.dart';

class AiFeatureSettingsRepository {
  final db.AppDatabase _db;

  AiFeatureSettingsRepository(this._db);

  Future<int> _activeProfileId() async {
    final row = await (_db.select(_db.appSettings)
          ..where((t) => t.key.equals('currentProfileId')))
        .getSingleOrNull();
    return int.tryParse(row?.value ?? '1') ?? 1;
  }

  Future<Map<String, dynamic>?> getSettings() async {
    final profileId = await _activeProfileId();
    final result = await _db.customSelect(
      'SELECT * FROM ai_feature_settings WHERE profile_id = ? LIMIT 1',
      variables: [drift.Variable(profileId)],
    ).getSingleOrNull();
    return result?.data;
  }

  Stream<Map<String, dynamic>?> watchSettings() {
    final profileStream = _db.select(_db.appSettings)..where((t) => t.key.equals('currentProfileId'));
    return profileStream
        .watchSingleOrNull()
        .map((row) => int.tryParse(row?.value ?? '1') ?? 1)
        .distinct()
        .asyncExpand((profileId) => _db
            .customSelect('SELECT * FROM ai_feature_settings WHERE profile_id = ? LIMIT 1',
                variables: [drift.Variable(profileId)], readsFrom: {_db.aiFeatureSettings})
            .watch()
            .map((rows) => rows.isEmpty ? null : rows.first.data));
  }

  /// Merge partial updates with the existing row so toggles do not NULL out other columns.
  Future<void> upsertSettings(Map<String, Object?> values) async {
    if (values.isEmpty) return;
    final profileId = await _activeProfileId();
    final existing = await (_db.select(_db.aiFeatureSettings)..where((t) => t.profileId.equals(profileId))).getSingleOrNull();

    bool merged(String snakeKey, bool prior) {
      if (!values.containsKey(snakeKey)) return prior;
      return _coerceBool(values[snakeKey], prior);
    }

    int mergedPositiveHours(String snakeKey, int prior) {
      if (!values.containsKey(snakeKey)) return prior;
      final v = values[snakeKey];
      if (v is int) return v > 0 ? v : prior;
      if (v is num) {
        final n = v.toInt();
        return n > 0 ? n : prior;
      }
      if (v is String) {
        final n = int.tryParse(v);
        if (n != null && n > 0) return n;
      }
      return prior;
    }

    final next = db.AiFeatureSetting(
      profileId: profileId,
      featChallengeAi: merged('feat_challenge_ai', existing?.featChallengeAi ?? false),
      featSessionInsights: merged('feat_session_insights', existing?.featSessionInsights ?? false),
      featStudyPlanner: merged('feat_study_planner', existing?.featStudyPlanner ?? false),
      featMotivation: merged('feat_motivation', existing?.featMotivation ?? false),
      featWeeklyReview: merged('feat_weekly_review', existing?.featWeeklyReview ?? false),
      coachEnabled: merged('coach_enabled', existing?.coachEnabled ?? false),
      smartChallengesEnabled: merged('smart_challenges_enabled', existing?.smartChallengesEnabled ?? false),
      debriefEnabled: merged('debrief_enabled', existing?.debriefEnabled ?? false),
      weeklyNarrativeEnabled: merged('weekly_narrative_enabled', existing?.weeklyNarrativeEnabled ?? false),
      subjectDifficultyEnabled: merged('subject_difficulty_enabled', existing?.subjectDifficultyEnabled ?? false),
      surpriseNotificationsEnabled:
          merged('surprise_notifications_enabled', existing?.surpriseNotificationsEnabled ?? false),
      surpriseCheckIntervalHours: mergedPositiveHours(
        'surprise_check_interval_hours',
        existing?.surpriseCheckIntervalHours ?? 3,
      ),
      updatedAt: DateTime.now().toUtc().toIso8601String(),
    );

    await _db.into(_db.aiFeatureSettings).insertOnConflictUpdate(
      db.AiFeatureSettingsCompanion(
        profileId: drift.Value(next.profileId),
        featChallengeAi: drift.Value(next.featChallengeAi),
        featSessionInsights: drift.Value(next.featSessionInsights),
        featStudyPlanner: drift.Value(next.featStudyPlanner),
        featMotivation: drift.Value(next.featMotivation),
        featWeeklyReview: drift.Value(next.featWeeklyReview),
        coachEnabled: drift.Value(next.coachEnabled),
        smartChallengesEnabled: drift.Value(next.smartChallengesEnabled),
        debriefEnabled: drift.Value(next.debriefEnabled),
        weeklyNarrativeEnabled: drift.Value(next.weeklyNarrativeEnabled),
        subjectDifficultyEnabled: drift.Value(next.subjectDifficultyEnabled),
        surpriseNotificationsEnabled: drift.Value(next.surpriseNotificationsEnabled),
        surpriseCheckIntervalHours: drift.Value(next.surpriseCheckIntervalHours),
        updatedAt: drift.Value(next.updatedAt),
      ),
    );
  }

  static bool _coerceBool(Object? v, bool fallback) {
    if (v == null) return fallback;
    if (v is bool) return v;
    if (v is int) return v != 0;
    if (v is num) return v != 0;
    if (v is String) {
      final s = v.toLowerCase();
      if (s == 'true' || s == '1') return true;
      if (s == 'false' || s == '0') return false;
    }
    return fallback;
  }
}

final aiFeatureSettingsRepositoryProvider = Provider<AiFeatureSettingsRepository>(
  (ref) => AiFeatureSettingsRepository(ref.watch(databaseProvider)),
);

File: src/core/data/repositories/aiFeatureSettingsRepository.ts
import { getDatabase } from '../database';

export type AiFeatureSettingsRow = {
  profileId: number;
  /** Legacy mirrors — synced on write for SQLite columns still present from v11 */
  featChallengeAi: boolean;
  featSessionInsights: boolean;
  featStudyPlanner: boolean;
  featMotivation: boolean;
  featWeeklyReview: boolean;
  coachEnabled: boolean;
  smartChallengesEnabled: boolean;
  debriefEnabled: boolean;
  weeklyNarrativeEnabled: boolean;
  subjectDifficultyEnabled: boolean;
  surpriseNotificationsEnabled: boolean;
  surpriseCheckIntervalHours: number;
  updatedAt: string;
};

type Row = {
  profile_id: number;
  feat_challenge_ai: number;
  feat_session_insights: number;
  feat_study_planner: number;
  feat_motivation: number;
  feat_weekly_review: number;
  coach_enabled: number | null;
  smart_challenges_enabled: number | null;
  debrief_enabled: number | null;
  weekly_narrative_enabled: number | null;
  subject_difficulty_enabled: number | null;
  surprise_notifications_enabled: number | null;
  surprise_check_interval_hours: number | null;
  updated_at: string;
};

function toBool(n: number | null): boolean {
  return n === 1;
}

function mapRow(r: Row): AiFeatureSettingsRow {
  const coach =
    r.coach_enabled !== null ? toBool(r.coach_enabled) : toBool(r.feat_motivation);
  const smart =
    r.smart_challenges_enabled !== null
      ? toBool(r.smart_challenges_enabled)
      : toBool(r.feat_challenge_ai);
  const debrief =
    r.debrief_enabled !== null ? toBool(r.debrief_enabled) : toBool(r.feat_session_insights);
  const weeklyNar =
    r.weekly_narrative_enabled !== null
      ? toBool(r.weekly_narrative_enabled)
      : toBool(r.feat_weekly_review);
  const subjDiff =
    r.subject_difficulty_enabled !== null
      ? toBool(r.subject_difficulty_enabled)
      : toBool(r.feat_study_planner);

  const surpriseEnabled =
    r.surprise_notifications_enabled !== null ? toBool(r.surprise_notifications_enabled) : false;
  const surpriseHoursRaw = r.surprise_check_interval_hours;
  const surpriseHours =
    surpriseHoursRaw != null && Number.isFinite(Number(surpriseHoursRaw)) && Number(surpriseHoursRaw) > 0
      ? Number(surpriseHoursRaw)
      : 3;

  return {
    profileId: r.profile_id,
    featChallengeAi: smart,
    featSessionInsights: debrief,
    featStudyPlanner: subjDiff,
    featMotivation: coach,
    featWeeklyReview: weeklyNar,
    coachEnabled: coach,
    smartChallengesEnabled: smart,
    debriefEnabled: debrief,
    weeklyNarrativeEnabled: weeklyNar,
    subjectDifficultyEnabled: subjDiff,
    surpriseNotificationsEnabled: surpriseEnabled,
    surpriseCheckIntervalHours: surpriseHours,
    updatedAt: r.updated_at
  };
}

export function defaultAiFeaturesForProfile(profileId: number): AiFeatureSettingsRow {
  const now = new Date().toISOString();
  return {
    profileId,
    featChallengeAi: false,
    featSessionInsights: false,
    featStudyPlanner: false,
    featMotivation: false,
    featWeeklyReview: false,
    coachEnabled: false,
    smartChallengesEnabled: false,
    debriefEnabled: false,
    weeklyNarrativeEnabled: false,
    subjectDifficultyEnabled: false,
    surpriseNotificationsEnabled: false,
    surpriseCheckIntervalHours: 3,
    updatedAt: now
  };
}

export function getAiFeaturesOrDefault(
  profileId: number,
  row: AiFeatureSettingsRow | null
): AiFeatureSettingsRow {
  return row ?? defaultAiFeaturesForProfile(profileId);
}

export async function getForProfile(profileId: number): Promise<AiFeatureSettingsRow | null> {
  const database = await getDatabase();
  const rows = await database.select<Row[]>(
    'SELECT * FROM ai_feature_settings WHERE profile_id = ? LIMIT 1',
    [profileId]
  );
  return rows[0] ? mapRow(rows[0]) : null;
}

export async function upsert(row: AiFeatureSettingsRow): Promise<void> {
  const database = await getDatabase();
  const b = (x: boolean) => (x ? 1 : 0);
  await database.execute(
    `INSERT INTO ai_feature_settings (
      profile_id,
      feat_challenge_ai, feat_session_insights, feat_study_planner, feat_motivation, feat_weekly_review,
      coach_enabled, smart_challenges_enabled, debrief_enabled, weekly_narrative_enabled, subject_difficulty_enabled,
      surprise_notifications_enabled, surprise_check_interval_hours,
      updated_at
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
    ON CONFLICT(profile_id) DO UPDATE SET
      feat_challenge_ai = excluded.feat_challenge_ai,
      feat_session_insights = excluded.feat_session_insights,
      feat_study_planner = excluded.feat_study_planner,
      feat_motivation = excluded.feat_motivation,
      feat_weekly_review = excluded.feat_weekly_review,
      coach_enabled = excluded.coach_enabled,
      smart_challenges_enabled = excluded.smart_challenges_enabled,
      debrief_enabled = excluded.debrief_enabled,
      weekly_narrative_enabled = excluded.weekly_narrative_enabled,
      subject_difficulty_enabled = excluded.subject_difficulty_enabled,
      surprise_notifications_enabled = excluded.surprise_notifications_enabled,
      surprise_check_interval_hours = excluded.surprise_check_interval_hours,
      updated_at = CURRENT_TIMESTAMP`,
    [
      row.profileId,
      b(row.smartChallengesEnabled),
      b(row.debriefEnabled),
      b(row.subjectDifficultyEnabled),
      b(row.coachEnabled),
      b(row.weeklyNarrativeEnabled),
      b(row.coachEnabled),
      b(row.smartChallengesEnabled),
      b(row.debriefEnabled),
      b(row.weeklyNarrativeEnabled),
      b(row.subjectDifficultyEnabled),
      b(row.surpriseNotificationsEnabled),
      row.surpriseCheckIntervalHours
    ]
  );
}
